src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -273,18 +273,111 @@ public void Shuffle<T>(T[] values)
         /// </remarks>
         public void Shuffle<T>(Span<T> values)
         {
-            int n = values.Length;
+            var chooser = new IncreasingUniform(this);
+            for (int i = 1; i < values.Length; i++)
+            {
+                int index = chooser.NextIndex();
+                (values[i], values[index]) = (values[index], values[i]);
+            }
+        }
+
+        private ref struct IncreasingUniform
+        {
+            private readonly Random _random;
+            private int _n;
+            private int _chunk;
+            private int _chunkRemaining;
 
-            for (int i = 0; i < n - 1; i++)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public IncreasingUniform(Random random)
             {
-                int j = Next(i, n);
+            _random = random;
+            _chunk = 0;
+            _n = 1;
+            _chunkRemaining = 0;
+            }
 
-                if (j != i)
+            /// <summary>
+            /// Increase n by one and then return a random positive integer less than the new n
+            /// </summary>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public int NextIndex()
+            {
+                int nextN = _n + 1;
+                if (_chunkRemaining == 0)
                 {
-                    T temp = values[i];
-                    values[i] = values[j];
-                    values[j] = temp;
+                    (int bound, int remaining) = CalculateBound(nextN);
+                    _chunkRemaining = remaining - 1;
+                    _chunk = _random.Next(bound);
                 }
+                else
+                {
+                    _chunkRemaining -= 1;
+                }
+
+                int result;
+                if (_chunkRemaining == 0)
+                {
+                    result = _chunk;
+                }
+                else
+                {
+                    result = (_chunk % nextN);
+                    _chunk /= nextN;
+                }
+
+
+                _n = nextN;
+                return result;
+            }
+
+            /// <summary>
+            /// Calculate the highest (x,k) such that x = n*n+1*..*n+k-1 where x is a 32 bit integer.
+            /// </summary>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static (int, int) CalculateBound(int n)
+            {
+                int count;
+                switch (n)
+                {
+                    case 2:
+                        return (479001600, 11); //12 factorial
```

Comment:
would it be worth adding some debug asserts to calculate and self document these magic numbers?